public/app/features/dashboard-scene/scene/DashboardScene.tsx

```diff
@@ -213,10 +217,37 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> impleme
     this.addActivationHandler(() => this._activationHandler());
   }
 
+  // Setups a listener to the url and saves the current query params if the url is the url of this dashboard scene.
+  private setupUrlListener = () => {
+    // return a noop function if the feature toggle is disabled
+    if (!config.featureToggles.preserveDashboardStateWhenNavigating) {
+      return () => {};
+    }
+
+    const unlisten = locationService.getHistory().listen((a) => {
+      if (this.state.uid && a.pathname.includes(this.state.uid)) {
+        this._currentQueryParams = a.search;
+      }
+    });
```

Comment:
directly subscribing to history is usually a bad idea (vs using react useLocation). History events happen before route changes. It's much better to use useLocation + effect  to update state. This way you do not need to check if the route is relevant for the current dashboard / route (which can be error prone).

https://github.com/grafana/scenes/blob/main/packages/scenes/src/services/useUrlSync.ts#L9

Reply:
@dprokop not sure I follow, why do you need the event before the route change? It looks like you ignore the history event when uid does not match. So I am confused why you need it here vs useLocation (which DashboardSceneRenderer already use). 

> I.e. imagine we are on /d/<UID>?orgId=1&var-a=1 and we are navigating using i.e. a link to /d/<UID-1>. To preserve the location correctly, I need to capture var-a=1 before the url has updated, and this code does exactly that.

If you add this to DashboardSceneRenderer (in render code or in useEffect, does not matter as _currentQueryParams is not part of state that triggers react re-render). 

```
model.rememberSearchParams(location.search) 
```

Would only save the URL before the link click to another dashboard

Reply:
@torkelo but adding this to the render path or useEffect will cause the params to be remembered every time the page renders. What I want tho is to remember the state of the url when I navigate to another dashboard, so that another dashboard can restore the url state when it initializes. If we do it in the render path, then we are always restoring the "own" params, don't we?

Reply:
> but adding this to the render path or useEffect will cause the params to be remembered every time the page renders.

But your could will also update the _currentQueryParams everytime the URL changes. But subscribing directly to history will also subscribe to history events that are unrelated to the current dashboard, something useLocation fixes. 

> What I want tho is to remember the state of the url when I navigate to another dashboard

yes, that is what will happen. useLocation inside DashboardSceneRenderer will not return a location for different dashboard so no need to compare uid